### ISSUE TITLE
Localhost

### ISSUE THREAD

**Issue by rfoel** [usebruno/bruno#124]
Does bruno work on localhost? I keep getting the following error when I send a request:

```
Error invoking remote method 'send-http-request': Error: connect ECONNREFUSED 127.0.0.1:3000
```

PS: My local server is running, I can call it with Insomnia, but not with bruno. 

I'm on a Mac M1.

**Comment by helloanoop** [comment]
@rfoel Apologies for checking on this late.

I was able to hit localhost on my machine.
I am using a Macbook Air and the bruno version is `v0.10.1`
![image](https://user-images.githubusercontent.com/9350733/226169564-f00fffb3-3971-4422-8fa3-3d5184e9c8d2.png)

Can you confirm if you are still facing the same issue?
Which OS and which version of bruno are you currently on ?

**Comment by rfoel** [comment]
I have Bruno v0.10.2 installed and macOS Ventura 13.1. I recorded a video with the issue.

https://user-images.githubusercontent.com/19496473/226332899-67d3ff18-f3dd-457c-88b4-1ae029670aa8.mov

**Comment by helloanoop** [comment]
@rfoel Can you confirm that curl works
`curl -X POST  -d '{"email": "rfoelf@gmail.com"}' http://localhost:3000/otp/request`

**Comment by rfoel** [comment]
@helloanoop yes it works fine.

**Comment by helloanoop** [comment]
Hey @rfoel !

Hope you are doing great!

It has been sometime.
No one else from the community has reported this issue.
There seems to be something on your machine that is preventing access to localhost from Bruno.

Can you install the latest version from [website](https://www.usebruno.com/downloads) and check if the issue persists ?

**Comment by rfoel** [comment]
Same error. I'm not really sure what's happening. Postman and Insomnia work fine on my machine. 

```
Error invoking remote method 'send-http-request': Error: connect ECONNREFUSED 127.0.0.1:3000
```

**Comment by fabioduque** [comment]
@rfoel Did you try to check if you have any firewall rule specific for Bruno, which might be blocking it? 
If curl, Postman and Insomnia work, then it seems to be some rule only affecting Bruno.

After reading your issue, I wanted to test that, so on my MBP M1, I've blocked Bruno using LuLu and I got the same error you have.

Maybe you can also try to remove the app and reinstall it.

**Comment by fabioduque** [comment]
Ok, sorry, disregard the part about LuLu. 

It is not blocking my connection to localhost after all. 

After setting the rule, I've closed Bruno, and when I opened it again, it didn't select my correct environment so my {{baseUrl}} was not being populated, hence the connection error. Sorry about that.

LuLu affects an external connection though...

**Comment by julianhirt** [comment]
I do have the same issue not using LuLu

**Comment by rfoel** [comment]
I've made some discoveries. In a project running an express app with docker, bruno seems to work with localhost. My problem still persists when I use `serverless-offline` though, so the problem might be something between this package and bruno.

**Comment by Epick362** [comment]
Experiencing the same with `serverless-offline` and Bruno :( 

```json
{
    "serverless": "^3.35.2",
    "serverless-offline": "13.2.0"
}
```

**Comment by julianhirt** [comment]
Personal update: My issue disappeared after upgrading to Mac OSX Sonoma. Can't reproduce it anymore.

**Comment by rfoel** [comment]
I guess the issue is specific with `serverless-offline` honestly, using different servers I wasn't able to reproduce the error.

**Comment by fabioduque** [comment]
@rfoel I'm not using `serverless-offline` and also had that issue, so it's not tied to that explicitly. 

I am just using Nuxt 3 (Vue 3), which starts everything with `vite`.
It might be something in common, of course.

Luckily, I have updated Bruno and my issue disappeared (meaning that I don't need to use `--host` option when starting Nuxt).

Btw, concerning Mac OSX Sonoma that @julianhirt mentioned, I had already upgraded when I had the issue with Bruno, so at least for me, Sonoma didn't solve it.

**Comment by jrc03c** [comment]
Hey, all! I'm on Linux Mint 21.2 (which I think is built on Ubuntu 22.04 LTS), and I'm seeing a similar error, except mine's happening on port 80:

![](https://i.ibb.co/1z41sPs/bruno-error.png)

I haven't modified any configurations or preferences or anything. I literally just installed Bruno a few minutes ago and tried making my first request — and this was the first thing I saw. (And I tried it several more different ways and got the same result each time.)

**Comment by helloanoop** [comment]
@jrc03c @fabioduque Is the issue unique to localhost or is it happening for public apis too?
We will be introducing logging very soon that will write stack traces into a file. Hoping that gets us closer to the root cause of the issue.

**Comment by fabioduque** [comment]
@helloanoop for me it was only happening on loopback (localhost or 127.0.0.1)
As soon as I added the `--host` flag which makes it accessible via IP (192.168.1.x, for example), it worked.

And also remember that for me, in the beginning it worked even without the `--host`.
Then suddenly it stopped working and after I updated Bruno, it started to work again! 

Super weird and I know that I'm not being very helpful with this info, because you won't be able to replicate or debug...
If it happens again, I'll share more info.

**Comment by ajcarrillo** [comment]
Hi. I'm getting the same error on Windows in both the installable and portable versions `v0.23.0`. I added the app to allow it in Windows firewall but that didn't work.

![image](https://github.com/usebruno/bruno/assets/4111721/1e81c2be-60c6-4002-af0f-f6240a3ba9ea)

**Comment by skyd0me** [comment]
Same for me.
Everything works fine with remote urls. If I use http://localhost:8080, I see the above mentioned error message.
Running on Windows 11 with v0.23.0 (portable version).

**Comment by austinwilcox** [comment]
So I am running into the same error as everyone here, in my case I think I know what the issue is, local ssl certs. I have set up local ssl certs with dotnet before on linux, but it was a bit of a bear. I will test that and report back if creating a self signed ssl cert will fix the issue with Bruno.

**Comment by helloanoop** [comment]
@austinwilcox Have you tried disabling SSL check in preferences..You can open preferences by clicking on the gear icon on the left bottom of the screen

**Comment by austinwilcox** [comment]
@helloanoop That fixed the issue for me! Thank you. Glad you got me before I dove into the bear of ssl certs and dotnet apps on linux.

**Comment by zarnautovic** [comment]
I have same issue with WSL2. my app is running on wsl2 and bruno is not able to make request. Postman is working fine. Someone know how to make it work ?

**Comment by rfoel** [comment]
> ok I could track it down to a wsl2 issue. My backend application is running within wsl2 and looks like bruno was not able to call this endpoint via http://localhost or http://127.0.0.1 But if I use http://[::1] --> IPv6 everything starts working. Need to check how I can adjust my wsl2 config, But it's still strange why this is an issue for bruno and not for postman etc. Maybe they have configured a fallback to IPv6, which is missing for bruno? Just a thought :-)

@helloanoop I really think @skyd0me comment is insightful, it is probably some IPv6 config that bruno is missing.

**Comment by Sleitnick** [comment]
Got this error when building out a small web server using sockets. I was using `AF_UNSPEC` for the `addrinfo.ai_family` field, which I guess was then choosing IPv6 down the line. Forced this to v4 by setting it to `AF_INET` instead, which "fixed" this issue for me. But yeah, I suspect something to do with IPv6, but I'm no expert with this sort of stuff.

**Comment by azivkovi** [comment]
Just to chip in, I get `Error invoking remote method 'send-http-request': Error: connect ECONNREFUSED 127.0.0.1:3000` using Bruno on Windows 11 and trying to reach my app on WSL2. Insomnia works.

**Comment by skyd0me** [comment]
I found a post which is saying that insomnia is doing an fallback implementation:
https://stackoverflow.com/questions/39358426/insomnia-rest-client-error

`The app will now first attempt an IPv6 lookup (usually ::1), then fallback to an IPv4 lookup (usually 127.0.0.1) on error`

and indeed. Looks like Curl is doing the same.
![image](https://github.com/usebruno/bruno/assets/18167556/572236b1-cd4e-4076-9b42-f9068cfeda15)

**Comment by zarnautovic** [comment]
I made it work with wsl2 changing http://localhost:port to http://[::1]:port. I can be quick fix until devs implement ipv6  lookup

**Comment by SandeepPoojari07** [comment]
Hello @helloanoop,

I have been facing the same issue on windows with Bruno v0.27.2
http://localhost:3000/ - doesn't work
http://[::1]:3000/ - works

Looking forward for a fix on the same.
Thanks.

**Comment by mkornblum** [comment]
Just wanted to add some more evidence in favor of the theory that something is off with IPv6 lookups. 

For local development, we run a few containers with Docker Compose behind a Traefik proxy, and they get hostnames like `http://my-service.localhost`. These are reachable via browsers (Chrome, Firefox), as well as via cURL, as well as via `fetch` from the dev tools console within Bruno itself. Looking in the Chrome dev tools, it resolves the URL to an IPv6 remote address like `[::1]:80`.

Trying to use the same hostname in the URL in the Bruno request UI results in the error:

 `Error invoking remote method 'send-http-request': Error: getaddrinfo ENOTFOUND my-service.localhost`
 
 Pulling the IPv4 URL assigned to the container (eg `http://192.168.107.11/`) off the Traefik dashboard, and using that in the Bruno UI instead, allows the requests to complete as expected.

**Comment by nomad483** [comment]
I have the same problem
server on yii2 (not seen on other technologies)
Postman, Curl - connects normally
Bruno - gives an error

OS: macOS Sonoma 14.1

**Comment by Fancman** [comment]
I am running website on apache2 in WSL2 WIndows 10. In hosts file I set custom local URLs for webapp and now when I am requesting API endpoints it is returning error which is getting everyone in thread. I cant fix it with ::1 in URL because I am running multiple websites in apache under 80 port. In Insomnia it works fine.

```shell
::1               www.oplatisa.sk.local
127.0.0.1         www.oplatisa.sk.local
::1               oplatisa.sk.local
127.0.0.1         oplatisa.sk.local
```

**Comment by skyd0me** [comment]
If I got it right, the concept to fallback to ipv4 is often referenced as happy eyeballs or RFC 8305.

Insomnia is using node-libcurl and there was an issue in the past with the same problems as mentioned here.
https://github.com/JCMais/node-libcurl/issues/280
This got resolved in the meantime and it looks like node-libcurl is doing the trick for them.

There is also a bigger PR discussion about implementing happy eyeballs in node 20: https://github.com/nodejs/node/pull/44731

Bruno is using axios. I was not able to find something related to happy eyeballs for axios.

**Comment by janos-r** [comment]
Same here,
Running on linux mint 21.2 with the .deb file from the official website (v1.1.1).
I would love to switch from Insomnia but not being able to query localhost is a deal breaker.

**Comment by skyd0me** [comment]
@janos-r:
have you tried the http://[::1]:PORT workaround?

**Comment by janos-r** [comment]
> @janos-r: have you tried the http://[::1]:PORT workaround?

Ok, it works now after adding `http://` at the front.

**Comment by notfounnd** [comment]
> I made it work with wsl2 changing http://localhost:port to http://[::1]:port. it can be quick fix until devs implement ipv6 lookup

Same behavior here. I've applied this tip and it works fine on localhost.

**Comment by azivkovi** [comment]
Feels like this issue should have a high priority as many people will be turned off by not being able to use localhost queries without a hack.

**Comment by notfounnd** [comment]
> Feels like this issue should have a high priority as many people will be turned off by not being able to use localhost queries without a hack.

I think the same, that was close that I not stay in the turned off list because this problem. I have returned to my Bruno's POC after resolve some pendencies here.

Hey @helloanoop, maybe this issue could need a fix/bug tag.

**Comment by BrandonGillis** [comment]
a coworker is having exactly the same issue on MacOS 13.2, Macbook Pro M2. The local env is running via docker and using traefik to forward request to the appropriate container, anyway with exactly the same docker stack on Linux, everything works fine

**Comment by BrandonGillis** [comment]
> @helloanoop Thanks for reaching out, but v1.3.2 still didnt fix it for my case. Website is running under WSL2 and in Insomnia I didnt change anything to make requests work. I am attaching screenshots of Bruno, Insomnia and proof that I am using right version.
> 
> ![image](https://private-user-images.githubusercontent.com/5480663/287980357-2098aba0-1b44-471a-b013-29da238b82a0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE3Njc2NTMsIm5iZiI6MTcwMTc2NzM1MywicGF0aCI6Ii81NDgwNjYzLzI4Nzk4MDM1Ny0yMDk4YWJhMC0xYjQ0LTQ3MWEtYjAxMy0yOWRhMjM4YjgyYTAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMDVUMDkwOTEzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDA5NjA2NWVlYWQ0ZWViMzliNTg3NzdlNjkyZGE1ODg4YmYxY2U4N2IxN2ZlMGJmNWRhMzdlYzEyMjY5NzBiNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.FlDht-4Vyz8xCE4a7k9dQsC_a_TN0LTv0huipk_1Nis) ![image](https://private-user-images.githubusercontent.com/5480663/287980468-4728e561-0bac-4dcf-99e3-a3a2e470c6f3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE3Njc2NTMsIm5iZiI6MTcwMTc2NzM1MywicGF0aCI6Ii81NDgwNjYzLzI4Nzk4MDQ2OC00NzI4ZTU2MS0wYmFjLTRkY2YtOTllMy1hM2EyZTQ3MGM2ZjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMDVUMDkwOTEzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODQ3OGY1MGNlMjZhZDk0MTNjYWY5OWZkZmNmMzQ3NzcxY2QxYmViNmRhNTczZTYwZDA4YWZkNWNjOGE5NTQxMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.rniR-1kc7-6uBm5g1HyCkD2ztvkWDlan6PYv4Twi4IY) ![image](https://private-user-images.githubusercontent.com/5480663/287980673-50b27554-461b-4e87-acbf-89acb593394f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE3Njc2NTMsIm5iZiI6MTcwMTc2NzM1MywicGF0aCI6Ii81NDgwNjYzLzI4Nzk4MDY3My01MGIyNzU1NC00NjFiLTRlODctYWNiZi04OWFjYjU5MzM5NGYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMDVUMDkwOTEzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NmJjNzI1NTU0YmYzMzc0NDQyYjIyNWM4ZTY1ODQ2MjBhY2Y2ZDEwZWFhODNjYTk4NWYzY2U3NjkxNzkxNmM5MiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.YL5BfhnxYLyfnSEK_edjyNebXBaF00-a9tMb0BTzlHQ)

Hello @Fancman, I see that your url is using *.local, unfortunately the fixes we've made are for *.localhost, that's why you're still encountering the issue, anyway we can easily add support for *.local. As a last test, would you mind doing the same test with Postman and see if it works? Thank you!

**Comment by Fancman** [comment]
> > @helloanoop Thanks for reaching out, but v1.3.2 still didnt fix it for my case. Website is running under WSL2 and in Insomnia I didnt change anything to make requests work. I am attaching screenshots of Bruno, Insomnia and proof that I am using right version.
> > ![image](https://private-user-images.githubusercontent.com/5480663/287980357-2098aba0-1b44-471a-b013-29da238b82a0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE3Njc2NTMsIm5iZiI6MTcwMTc2NzM1MywicGF0aCI6Ii81NDgwNjYzLzI4Nzk4MDM1Ny0yMDk4YWJhMC0xYjQ0LTQ3MWEtYjAxMy0yOWRhMjM4YjgyYTAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMDVUMDkwOTEzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDA5NjA2NWVlYWQ0ZWViMzliNTg3NzdlNjkyZGE1ODg4YmYxY2U4N2IxN2ZlMGJmNWRhMzdlYzEyMjY5NzBiNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.FlDht-4Vyz8xCE4a7k9dQsC_a_TN0LTv0huipk_1Nis) ![image](https://private-user-images.githubusercontent.com/5480663/287980468-4728e561-0bac-4dcf-99e3-a3a2e470c6f3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE3Njc2NTMsIm5iZiI6MTcwMTc2NzM1MywicGF0aCI6Ii81NDgwNjYzLzI4Nzk4MDQ2OC00NzI4ZTU2MS0wYmFjLTRkY2YtOTllMy1hM2EyZTQ3MGM2ZjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMDVUMDkwOTEzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODQ3OGY1MGNlMjZhZDk0MTNjYWY5OWZkZmNmMzQ3NzcxY2QxYmViNmRhNTczZTYwZDA4YWZkNWNjOGE5NTQxMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.rniR-1kc7-6uBm5g1HyCkD2ztvkWDlan6PYv4Twi4IY) ![image](https://private-user-images.githubusercontent.com/5480663/287980673-50b27554-461b-4e87-acbf-89acb593394f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE3Njc2NTMsIm5iZiI6MTcwMTc2NzM1MywicGF0aCI6Ii81NDgwNjYzLzI4Nzk4MDY3My01MGIyNzU1NC00NjFiLTRlODctYWNiZi04OWFjYjU5MzM5NGYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMDVUMDkwOTEzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NmJjNzI1NTU0YmYzMzc0NDQyYjIyNWM4ZTY1ODQ2MjBhY2Y2ZDEwZWFhODNjYTk4NWYzY2U3NjkxNzkxNmM5MiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.YL5BfhnxYLyfnSEK_edjyNebXBaF00-a9tMb0BTzlHQ)
> 
> Hello @Fancman, I see that your url is using *.local, unfortunately the fixes we've made are for *.localhost, that's why you're still encountering the issue, anyway we can easily add support for *.local. As a last test, would you mind doing the same test with Postman and see if it works? Thank you!

I changed website configuration on *.localhost and it works now. Thank you

**Comment by BrandonGillis** [comment]
@helloanoop I think we can add support for special TLD domain see RFC 2606, 2: https://www.rfc-editor.org/rfc/rfc2606#section-2

It would mean adding support for :
.test
.example
.invalid
.localhost

For .local TLD it's a little more complicated see RFC 6762, 3 : https://datatracker.ietf.org/doc/html/rfc6762#section-3

But I think it's safe to resolve to the loopback if there's no dns resolution from the host, what do you think?

**Comment by BrandonGillis** [comment]
> I am running website on apache2 in WSL2 WIndows 10. In hosts file I set custom local URLs for webapp and now when I am requesting API endpoints it is returning error which is getting everyone in thread. I cant fix it with ::1 in URL because I am running multiple websites in apache under 80 port. In Insomnia it works fine.
> 
> ```shell
> ::1               www.oplatisa.sk.local
> 127.0.0.1         www.oplatisa.sk.local
> ::1               oplatisa.sk.local
> 127.0.0.1         oplatisa.sk.local
> ```

@Fancman just saw that you had the domains in your hosts file, so it should have worked. I will try to reproduce the issue you had with *.local, could you share exactly how it's setup on your machine, for example :
- wsl2 run an http server on test.local on port 8080
- Bruno run on windows, but can't reach test.local
- hosts file was modified on windows system32/driver/etc/hosts or is it the host file of wsl2 in /etc/hosts?

Thanks a lot!

**Comment by Fancman** [comment]
> > I am running website on apache2 in WSL2 WIndows 10. In hosts file I set custom local URLs for webapp and now when I am requesting API endpoints it is returning error which is getting everyone in thread. I cant fix it with ::1 in URL because I am running multiple websites in apache under 80 port. In Insomnia it works fine.
> > ```shell
> > ::1               www.oplatisa.sk.local
> > 127.0.0.1         www.oplatisa.sk.local
> > ::1               oplatisa.sk.local
> > 127.0.0.1         oplatisa.sk.local
> > ```
> 
> @Fancman just saw that you had the domains in your hosts file, so it should have worked. I will try to reproduce the issue you had with *.local, could you share exactly how it's setup on your machine, for example :
> 
> * wsl2 run an http server on test.local on port 8080, do you bind ::1 explicitly?
> * Bruno run on windows, but can't reach test.local
> * hosts file was modified on windows system32/driver/etc/hosts or is it the host file of wsl2 in /etc/hosts?
> 
> Thanks a lot!

On WSL I installed apache2, created and enabled config for website which listens on 80 port:

#### **`/etc/apache2/sites-available/oplatisa.sk.conf`**
```bash
<VirtualHost *:80>
        ServerName www.oplatisa.sk.local
        ServerAlias oplatisa.sk.local
        DocumentRoot /var/www/oplatisa.sk/public
        ErrorLog ${APACHE_LOG_DIR}/error.log
        CustomLog ${APACHE_LOG_DIR}/access.log combined
        # To make URL segments work
        <Directory /var/www/oplatisa.sk/public>
                Options Indexes FollowSymLinks
                AllowOverride All
                Require all granted
        </Directory>
</VirtualHost>
```

Then I modified `C:\Windows\System32\drivers\etc\hosts` and it seems that it automatically got replicated into `/etc/hosts` because it contains same addresses.

#### **`C:\Windows\System32\drivers\etc\hosts`**
```bash
::1               www.oplatisa.sk.local
127.0.0.1         www.oplatisa.sk.local
::1               oplatisa.sk.local
127.0.0.1         oplatisa.sk.local

::1               www.oplatisa.sk.localhost
127.0.0.1         www.oplatisa.sk.localhost
::1               oplatisa.sk.localhost
127.0.0.1         oplatisa.sk.localhost
```

#### **`\etc\hosts`**
```bash
# This file was automatically generated by WSL. To stop automatic generation of this file, add the following entry to /etc/wsl># [network]
# generateHosts = false
127.0.0.1       localhost
127.0.1.1       nb-figura.      nb-figura

::1     www.oplatisa.sk.local
127.0.0.1       www.oplatisa.sk.local
::1     oplatisa.sk.local
127.0.0.1       oplatisa.sk.local
::1     www.oplatisa.sk.localhost
127.0.0.1       www.oplatisa.sk.localhost
::1     oplatisa.sk.localhost
127.0.0.1       oplatisa.sk.localhost

# The following lines are desirable for IPv6 capable hosts
::1     ip6-localhost ip6-loopback
fe00::0 ip6-localnet
ff00::0 ip6-mcastprefix
```

Then I tried to make requests from Bruno which is installed on Windows 11 and couldnt access it. Before that I had website running in WSL under localhost and when I tried to access it with Bruno from Windows it worked (GET 127.0.0.1). Problems started when I changed website address to oplatisa.sk.local

**Comment by BrandonGillis** [comment]
Thanks for the detailed answer, that will help a lot to reproduce the issue!

**Comment by emflomed17** [comment]
I believe for MacOS users this is still an issue right?

I'm still getting the: _"Error invoking remote method 'send-http-request': Error: connect ECONNREFUSED 127.0.0.1:3000"_

Request example: http://localhost:3000/dev/poc/1a9c4c0f-0ee2-4155-9848-1d047382c5a7

It is a serverless framework project.

My Bruno version is 1.7.1

I have:
M2 Chip
Sonoma 14.2.1 

Kind of sad because I really wanted to give it a shot to Bruno.

**Comment by anitachan** [comment]
I have the same issue. I'm using 1.6.1.  I have an intel mac with Sonoma 14.3. In insomnia works well but bruno doesn't. This is my error!  `Error invoking remote method 'send-http-request': Error: connect ECONNREFUSED ::1:8080`

**Comment by muffe** [comment]
Same issue here

macOS: 14.3.1
Bruno: 1.10.0

Trying to run a pre request script

```
const fetch = require("node-fetch");
const body = {
  email: "{{username}}",
  password: "{{password}}"
};

const response = await fetch('http:/127.0.0.1:3000/api/users/login', {
	method: 'post',
	body: JSON.stringify(body),
	headers: {'Content-Type': 'application/json'}
});
const data = await response.json();

bru.setVar("ACCESS_TOKEN", data.accessToken);
```

The error given is
```
Error invoking remote method 'send-http-request': 
FetchError: request to http://127.0.0.1:3000/api/users/login failed, reason: connect ECONNREFUSED 127.0.0.1:3000
```

The application is running at 127.0.0.1:300.
If I run a "normal" request in bruno against the endpoint it works, but as soon as the context is the pre request script the above error occurs.

**Comment by jquinc30** [comment]
For @druxton-consid and the ones who continue to experiment the issue, I found 2 workarounds for a teammate that both works for us but I really prefer the second one:

1. In the URL, replace `localhost` by your ip address [for example, you can find it thanks to the output of the command `ifconfig`]
   - Example: `http://localhost:3000/xyz` should be replaced by `http://yourIP:3000/xyz`. So if `yourIP` is `10.123.124.125`, it should be `http://10.123.124.125:3000/xyz`
   - NB: with this solution, your will need to change the URL each time your IP address updates. Check the second point for a more robust/long term solution.
2. Update your OS _hosts_ file (`/private/etc/hosts` on a mac) and create a fake local domain name to trick Bruno.
  - Add `127.0.0.1         bruno.local` [here `bruno.local` is arbitrary and can be replaced]  (cf line 12 in the below screenshot)
![image](https://github.com/usebruno/bruno/assets/50746914/cd6e39fd-58da-4176-8d60-bca9002d2259)
  - Replace `localhost` by this domain in URL in Bruno.
    - Example: `http://localhost:3000/xyz` should be replaced by `http://bruno.local:3000/xyz`.

**Comment by mig8447** [comment]
This is still an issue in macOS for me, I managed to fix it by using `http://127.0.0.1` instead of `http://localhost` but this is still an issue with bruno (even the CLI) preferring IPV6

**Comment by douglasmiranda** [comment]
Bruno 1.12.3
System: Pop!_OS 22.04 
Using Bruno AppImage version

Basically same thing happening with me.

Trying to access http API running in container. Ran into the same problems of localhost vs 127.0.0.1 and in some cases my request went through but I've noticed my request with a json body just wont  go through.

Doing the same exact comand with cURL or using httpie works just fine.

For reference, I tried the example in this API specifically https://docs.browserless.io/HTTP-APIs/pdf

**Comment by trevorrobertsjr** [comment]
Bruno 1.19.0
CPU: M3 Pro
OS: macOS 14.5

I am unable to access my flask[async] server via http://localhost. It works with http://127.0.0.1. It works fine with curl
If I use the standard wsgi flask server or with gunicorn, http://localhost works.

**Comment by Snafuh** [comment]
Same issue here
Bruno 1.24.0
Windows 10

The weird thing is it worked earlier this week. Opened my collection today and nothing worked. Browser and curl are working fine.
Using 127.0.0.1 does not work for me either.

**Comment by berain-uvic** [comment]
On Windows 11 I was having the same issue, uninstalling and reinstalling solved the isssue.

**Comment by krestaino** [comment]
On Windows 11 after a fresh install of Bruno and running it the first time, I was having this issue. Not the greatest first impression, but simply closing the program and opening it again resolved the issue.

**Comment by jeffreyrwells** [comment]
> This is still an issue in macOS for me, I managed to fix it by using `http://127.0.0.1` instead of `http://localhost` but this is still an issue with bruno (even the CLI) preferring IPV6
> 
> For the CLI you can use the following:
> 
> ```shell
> NODE_OPTIONS=--dns-result-order=ipv4first npx -y --package @usebruno/cli bru run
> ```
> 
> This will make the Bruno CLI to use ipv4 name resolution first

Changing http://localhost to http://127.0.0.1 fixed the Bruno CLI for my situation (CLI specific issue only). Thank you!
